### PR TITLE
Recompose slots

### PR DESCRIPTION
**Pull request by khmakoto**
#### Pull request checklist

- [x] Addresses an existing issue: Related to #8331
- [x] Include a change request file using `$ yarn change`

#### Description of changes

This PR allows the the `composed` handle to recompose already-created components by changing some of the options like `slots` or `view`.

#### Focus areas to test

Check the slots recomposition example under `experiments/src/slots/examples` and see recomposition at work:

image

Recomposition code for first Button:

```ts
export const SpinnerButton: React.StatelessComponent<IButtonProps> = composed<IButtonProps, IButtonTokens, IButtonStyles>(Button, {
  slots: { icon: Spinner }
});
```

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/OfficeDev/office-ui-fabric-react/pull/10345)

### CODE REVIEW

File: packages/experiments/src/slots/examples/Slots.Recomposition.Example.tsx [hunk 1]
```diff
@@ -0,0 +1,73 @@
+/** @jsx withSlots */
```
Comment (dzearing) [inline packages/experiments/src/slots/examples/Slots.Recomposition.Example.tsx]: withSlots no longer needed?

File: packages/experiments/src/slots/examples/Slots.Recomposition.Example.tsx [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+      onClick(ev);
+
+      if (ev.defaultPrevented) {
+        return;
+      }
+    }
+  };
+
+  return (
+    <slots.root
```
Comment (dzearing) [inline packages/experiments/src/slots/examples/Slots.Recomposition.Example.tsx]: I didn't think this was valid JSX; I thought that "slots" would need to be capitalized?

File: packages/foundation/src/next/composed.tsx [hunk 3]
```diff
@@ -42,8 +43,96 @@ export function composed<
...
+  >).__options;
+
+  // Check if we are composing or recomposing.
+  let options: IComponentOptions<TComponentProps, TTokens, TStyleSet, TViewProps, TComponentSlots, TStatics>;
+  if (typeof baseComponentOrOptions === 'function' && baseComponentOptions) {
+    let slots: ISlotComponent<TComponentProps, TComponentSlots>;
+    const baseComponentSlots = baseComponentOptions.slots ? baseComponentOptions.slots : ({} as ISlotDefinition<Required<TComponentSlots>>);
+    const recompositionSlots = recompositionOptions ? (recompositionOptions.slots ? recompositionOptions.slots : {}) : {};
+
+    if (typeof baseComponentSlots === 'function') {
```
Comment (dzearing) [inline packages/foundation/src/next/composed.tsx]: Why would `baseComponentSlots` be a function? It seems like you'd not want slots to be a function of props.

File: packages/foundation/src/next/composed.tsx [hunk 4]
```diff
@@ -42,8 +43,96 @@ export function composed<
...
+        slots = props => {
+          return { ...baseComponentSlots(props), ...recompositionSlots(props) };
+        };
+      } else {
+        slots = props => {
+          return { ...baseComponentSlots(props), ...recompositionSlots };
+        };
+      }
+    } else {
+      if (typeof recompositionSlots === 'function') {
```
Comment (dzearing) [inline packages/foundation/src/next/composed.tsx]: really having a hard time following this. Why the need to resolve functional slot collection?

File: packages/foundation/src/next/composed.tsx [hunk 5]
```diff
@@ -42,8 +43,96 @@ export function composed<
...
+  TViewProps extends TComponentProps = TComponentProps,
+  TComponentSlots = {},
+  TStatics = {}
+>(
+  baseComponentOrOptions:
+    | React.FunctionComponent
+    | IComponentOptions<TComponentProps, TTokens, TStyleSet, TViewProps, TComponentSlots, TStatics> = {},
+  recompositionOptions?: IRecompositionComponentOptions<TComponentProps, TTokens, TStyleSet, TViewProps, TComponentSlots, TStatics>
 ): React.FunctionComponent<TComponentProps> & TStatics {
+  const baseComponentOptions = (baseComponentOrOptions as ISlotCreator<
```
Comment (dzearing) [inline packages/foundation/src/next/composed.tsx]: oi; is this case needed? It really makes this harder to read.

File: packages/experiments/src/slots/examples/Slots.Recomposition.Example.tsx (hunk 2 above)
Comment (JasonGore) [reply packages/experiments/src/slots/examples/Slots.Recomposition.Example.tsx]: Capitalization shouldn't matter as long as it's the right type.

Comment (khmakoto) [reply packages/experiments/src/slots/examples/Slots.Recomposition.Example.tsx]: Well, this is just something that is being passed as a parameter into the view function. I can capitalize it if you want, the corresponding line is line 16.

File: packages/experiments/src/slots/examples/Slots.Recomposition.Example.tsx (hunk 1 above)
Comment (JasonGore) [reply packages/experiments/src/slots/examples/Slots.Recomposition.Example.tsx]: If there are no longer any closures (which I don't see) then we can finally get rid of withSlots. 🎉 Should also check the React hierarchy and make sure slotted components are not rerendering on every pass, as well.

Comment (khmakoto) [reply packages/experiments/src/slots/examples/Slots.Recomposition.Example.tsx]: I _think_ it is still needed, I remember the page throwing me an error in the new example I added when it didn't have `withSlots` at the top.

File: packages/foundation/src/next/composed.tsx (hunk 3 above)
Comment (khmakoto) [reply packages/foundation/src/next/composed.tsx]: What happens for something like `Button` where the default `root` slot can change from `button` to `a` when `href` is set?

File: packages/foundation/src/next/composed.tsx (hunk 4 above)
Comment (khmakoto) [reply packages/foundation/src/next/composed.tsx]: If slots can be functions then the resolution for `base` + `options` would need to take this into account, wouldn't it?

File: packages/experiments/src/slots/examples/Slots.Recomposition.Example.tsx (hunk 1 above)
Comment (JasonGore) [reply packages/experiments/src/slots/examples/Slots.Recomposition.Example.tsx]: Oh, I bet it's needed for `Stack`. Assuming `composed` is not using `slots.ts` (which I don't see) then none of the components you've created should be throwing the error.

Comment (JasonGore) [reply packages/experiments/src/slots/examples/Slots.Recomposition.Example.tsx]: Scratch that, it's definitely not needed for `Stack`. It should only be needed in files that render the old-style slots. Can you check again?

File: packages/foundation/src/next/composed.tsx (hunk 5 above)
Comment (khmakoto) [reply packages/foundation/src/next/composed.tsx]: I think there are ways in which we can make this less cumbersome, I'll experiment.

File: packages/experiments/src/slots/examples/Slots.Recomposition.Example.tsx (hunk 1 above)
Comment (khmakoto) [reply packages/experiments/src/slots/examples/Slots.Recomposition.Example.tsx]: For sure, I'll check again.

Comment (khmakoto) [reply packages/experiments/src/slots/examples/Slots.Recomposition.Example.tsx]: By the way, I think `composed` is still using `slots.ts`, especifically the functions `createFactory` and `getSlots`. Just wanted to point that out.

Comment (JasonGore) [reply packages/experiments/src/slots/examples/Slots.Recomposition.Example.tsx]: Ok, then `getSlots` is what's throwing the error. I wasn't sure if we'd be able to get rid of the closures yet, but as David's mentioned, that will change how we use slots and require the view to spread slotprops manually on each slot.

File: packages/foundation/src/next/composed.tsx (hunk 3 above)
Comment (dzearing) [reply packages/foundation/src/next/composed.tsx]:
```
const resolveWith = (thing, data) => typeof thing === 'function' ? thing(data) : thing;

slots = props => {
  ...resolveWith(baseComponentSlots, props),
  ...resolveWith(recompositionSlots, props)
};
```

File: packages/foundation/src/next/composed.tsx (hunk 4 above)
Comment (dzearing) [reply packages/foundation/src/next/composed.tsx]: nevermind!

**Pull request merged**